### ISSUE TITLE
ExecutionEngineException on using step into while debugging

### ISSUE THREAD

**Issue by aromaa** [dotnet/runtime#95417]
### Description

When trying to step into a virtual call the runtime unexpectedly throws ExecutionEngineException. Setting the `<TieredCompilation>false</TieredCompilation>` flag to the project file seems to fix this.

### Reproduction Steps

Setting a breakpoint on the following code on L9 and executing step into the 6th time would trigger ExecutionEngineException when attempting to step into the `IDbContextDependencies.StateManager` on `DbContext`.

```csharp
using Microsoft.EntityFrameworkCore;

using TestContext context = new();

TestEntity entity = new();

_ = context.Entities.Entry(entity);

context.Entities.Update(entity); //Start executing step into in this line.

internal sealed record TestEntity(int Id = 0);

internal sealed class TestContext : DbContext
{
	public DbSet<TestEntity> Entities => this.Set<TestEntity>();

	protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder) =>
		optionsBuilder.UseInMemoryDatabase("test");
}
```

### Expected behavior

No exception

### Actual behavior

The runtime crashes with ExecutionEngineException

### Configuration

.NET 8
Windows 10 22H2 build 19045.3693
x64
Microsoft Visual Studio Community 2022 Version 17.9.0 Preview 1.0

**Comment by MichalPetryka** [comment]
Does `<TieredPGO>false</TieredPGO>` help too?

**Comment by aromaa** [comment]
> Does `<TieredPGO>false</TieredPGO>` help too?

It does not.

**Comment by aromaa** [comment]
[Dump](https://drive.google.com/file/d/1JPY4POx12rFO3DzQVy6IJ34RIbq4q41y/view?usp=sharing)

I was unable to take one from VS directly so I used the task manager instead.

**Comment by tommcdon** [comment]
HI @aromaa I apologize for the delay and have requested access to the dump to investigate.

**Comment by aromaa** [comment]
You should be able to access it now.

**Comment by mikelle-rogers** [comment]
@aromaa, just clarifying that by line 9 you mean this line? `protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder) =>
		optionsBuilder.UseInMemoryDatabase("test");`

**Comment by aromaa** [comment]
No, I mean literally the L9. The one with the `context.Entities.Update(entity);`.

**Comment by tommcdon** [comment]
Hi @aromaa! I'm sorry for the long delay in getting back to you on this issue. I am still not able to reproduce the issue, on .NET 8, 9, or current main. The dump indicates that the instruction pointer inside the context record was somehow corrupted. I traced through the code path in the debugger and didn't see any place where we could be resetting it. The odd thing is that the ExceptionRecord has the correct IP, but RIP in the ContextRecord is incorrect, which indicates something overwrote RIP in the context record either before or after the Vectored Exception Handler callback in the runtime. I'm hoping you can answer some more clarifying questions to help us understand the issue better. Are you able to see this problem occurring on multiple computers, or does this only occur on one machine? Do you see the same issue with other applications on online this repro? Does this reproduce on .NET framework?

**Comment by aromaa** [comment]
At the time I did ask my friend whatever it crashed on their computer, it did.

This does not seem to repro anymore on 8.0.18 but reverting back to 8.0.0 by specifying `<RollForward>Disable</RollForward>` and `<RuntimeFrameworkVersion>8.0.0</RuntimeFrameworkVersion>` does.

Presumably some patches have been applied to the relevant logic, given the issue age. I think we can close?

**Comment by tommcdon** [comment]
Thanks for the quick response @aromaa! I was testing on the latest .NET 8 and 9, and didn't test against 8.0.0. We will move forward with closing as you were able to reproduce the problem on 8.0.0 and then validate that updating fixed the issue. Thanks!
